src/sql/src/plan/query.rs

```diff
@@ -5378,17 +5378,7 @@ pub fn scalar_type_from_catalog(
                 CatalogType::Record { fields } => {
                     let scalars: Vec<(ColumnName, ColumnType)> = fields
                         .iter()
-                        .map(|f| {
-                            let scalar_type =
-                                scalar_type_from_catalog(scx, f.type_reference, &f.type_modifiers)?;
```

Comment:
I don't understand why this line move from here to plan_create_type. Presumably it needed to be here for some reason? Was it previously cargo culted and could have been in plan_create_type all along? I'm of the belief that the current TypeReference generic exists for a reason, although I'm not sure what it is. Are there problems keeping this as written?

Reply:
To return the `RelationDesc` of a composite type from it's `fields: Vec<CatalogRecordField>`, we'll need the scalar_type along with the modifiers. The modifiers are already there in a `CatalogRecordField`, but we only have the `GlobalId` of the type via this `type_reference`. From what I understood, the only way to get a scalar type from a GlobalId is to query it from catalog via `scalar_type_from_catalog`, which we don't have access to in the `desc` call. So, I stored it in the `CatalogRecordField`.

One issue I could see with this is with deeply nested types the scalar type object could become big vs just keeping a reference. I don't know how bad it would be though.

Let me think about this a bit more if it can be refactored differently. Also, let me know if you can think of a different way to do this.

Reply:
I checked all the callsites of `desc`. At each one of them the catalog is in scope. Maybe I can pass the catalog as an argument to look up the scalar type only while creating the `RelationDesc`?

Reply:
Sadly this code is not well understood by our current team. I'm going to do some investigation myself to research exactly why TypeReference exists, then can hopefully give an informed opinion.

Reply:
It's to allow constructing built-in types in the catalog that reference other built-in types before those types have IDs. The reason that's not biting us here at the moment is because there aren't any built-in types that are record types at present. But there might be someday!

Philosophically, `CatalogType` lives between the PostgreSQL domain (`mz_pgrepr::Type`) and the Materialize domain (`ColumnType` and `ScalarType`). It represents an entry in `mz_types`, essentially, and is what bridges the gap between PostgreSQL world and Materialize world. It intentionally doesn't contain `ColumnType` or `ScalarType`, as those are lower level concepts, but it's meant to be easily convertible to a `ColumnType` given access to a catalog to resolve the references.

The trick here is to cache the `desc` on `catalog::Type`. That way it's available to the `CatalogEntry::desc()` method without requiring the catalog to be passed in to `desc`, but also we keep `CatalogType` consistent with its philosophy (no containing `ColumnType`s).

This is all massively confusing, for sure. Sorry. Things in the type system developed rather organically, and while most of the concepts are fundamentally sound, the naming (`ColumnType` vs `CatalogType` vs `catalog::Type`) really leaves something to be desired.

Reply:
I pushed up a commit demonstrating the `desc` trick.

Reply:
Oh nice. We are keeping the `desc` during creation and when parsing itself. So, still keeping the ID reference, won't blow up in case there's nested records.